### PR TITLE
[kimi2.7] pass correct fsdp mesh for backend=spmd_types

### PR DESCRIPTION
**Pull request by weifengpy**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #4060
* __->__ #4063

before: spmd_types exposes the dense DP axis as dp_shard, but Kimi 2.7 requested the legacy fsdp axis, causing Invalid mesh dim: 'fsdp' during initialization

after: explicitly tells FSDP that dp_shard or efsdp is the data-parallel shard axis

### CODE REVIEW

File: torchtitan/models/qwen3_5/parallelize.py [hunk 1]
```diff
@@ -96,6 +96,7 @@ def parallelize_qwen3_5(
             reduce_dtype=TORCH_DTYPE_MAP[training.mixed_precision_reduce],
             reshard_after_forward_policy=parallelism.fsdp_reshard_after_forward,
             pp_enabled=parallel_dims.pp_enabled,
+            dp_mesh_dims=None,
```
Comment (shuhuayu) [inline torchtitan/models/qwen3_5/parallelize.py]: why this change, maybe we can do this in a separate commit for qwen 3.5?

File: torchtitan/distributed/fsdp.py [hunk 2]
```diff
@@ -85,6 +85,8 @@ def apply_fsdp_to_vision_encoder(
     reduce_dtype: torch.dtype,
     reshard_after_forward_policy: str = "default",
     pp_enabled: bool = False,
+    *,
+    dp_mesh_dims: DataParallelMeshDims | None,
```
Comment (shuhuayu) [inline torchtitan/distributed/fsdp.py]: can we make it optional to be consistent with `apply_fsdp_to_decoder`, i.e., use default value `None`?

**Review by shuhuayu — CHANGES_REQUESTED** [review]
thanks for adding this. i did a debug test run and hit some errors, related to model.parallelize(). llama3 has done this check: https://github.com/pytorch/torchtitan/blob/b175497ea3502388bb9513391810a86ca46dbfa8/torchtitan/models/llama3/parallelize.py#L43
can you do a test run for the debug variant to check further?

File: torchtitan/models/qwen3_5/parallelize.py (hunk 1 above)
Comment (weifengpy) [reply torchtitan/models/qwen3_5/parallelize.py]: from below, seems you are comfortable with `dp_mesh_dims: DataParallelMeshDims | None`, I can delete this qwen change

File: torchtitan/models/kimi_k2_7/sharding.py [hunk 3]
```diff
@@ -112,6 +113,7 @@ def _set_vision_encoder_sharding(ve_cfg) -> None:
         state_shardings={"weight": _REPLICATE_PARAM, "bias": _REPLICATE_PARAM},
         in_src_shardings={"input": _REPLICATE_ACT},
         in_dst_shardings={"input": _REPLICATE_ACT},
+        out_src_shardings=_REPLICATE_ACT,
```
Comment (weifengpy) [inline torchtitan/models/kimi_k2_7/sharding.py:116]: required once model.parallelize() runs under DP-only spmd_types

**Comment by weifengpy** [comment]
> can you do a test run for the debug variant to check further?

fixed. added integration test to catch this but let me know if you want such test

File: torchtitan/distributed/fsdp.py (hunk 2 above)
Comment (weifengpy) [reply torchtitan/distributed/fsdp.py]: done

**Review by shuhuayu — APPROVED** [review]
lgtm! maybe @pianpwk can have a final check about the spmd type checks, is multimodal tensor check solved now?

**Comment by pianpwk** [comment]
> lgtm! maybe @pianpwk can have a final check about the spmd type checks, is multimodal tensor check solved now?

hmm, does this pass with spmd typechecking? or is it just enabling the spmd backend. I can take the typechecking portion if not.

**Comment by shuhuayu** [comment]
> > lgtm! maybe @pianpwk can have a final check about the spmd type checks, is multimodal tensor check solved now?
> 
> hmm, does this pass with spmd typechecking? or is it just enabling the spmd backend. I can take the typechecking portion if not.

no, spmd type checking is deliberately ignored, see here: https://github.com/pytorch/torchtitan/blob/bcc09297a1d892e0ccf6a41fc367e76aab02475b/tests/integration_tests/models.py#L15

so this is only for enablement.

File: tests/integration_tests/models.py [hunk 4]
```diff
@@ -240,6 +240,20 @@ def build_model_tests_list() -> list[OverrideDefinitions]:
             ngpu=8,
         ),
         # Integration Test Cases for Kimi K2.7
+        OverrideDefinitions(
+            [
+                [
+                    # Do not enable --debug.spmd_typechecking: multimodal pixel
```
Comment (weifengpy) [inline tests/integration_tests/models.py:246]: @pianpwk sorry, no spmd_typechecking this time

**Comment by weifengpy** [comment]
oops, this wasn't merging into main
@shuhuayu I created a new PR with the same change for main, but need your approval: https://github.com/pytorch/torchtitan/pull/4070
